src/views/projects/SourceBrowser/FileDiff.svelte

```diff
@@ -58,6 +72,101 @@
       }
     }
   }
+
+  function thisLineSelected(
+    uriHunkLine: string | undefined,
+    hunkIdx: number,
+    lineIdx: number,
+  ): boolean {
+    if (!isSelected || !uriHunkLine) {
+      return false;
+    }
+
+    // range or single selection
+    const match = uriHunkLine.match(/H(\d*)L(\d*)L?(\d*)/);
+    if (match) {
+      if (match[3]) {
+        return (
+          match[1] === hunkIdx.toString() &&
+          lineIdx >= parseInt(match[2]) &&
+          lineIdx <= parseInt(match[3])
+        );
+      } else {
+        return uriHunkLine === "H" + hunkIdx + "L" + lineIdx;
+      }
+    }
+
+    return false;
+  }
+
+  function shouldScroll(
+    uriHunkLine: string | undefined,
+    hunkIdx: number,
+    lineIdx: number,
+  ): boolean {
+    if (!isSelected || !uriHunkLine) {
+      return false;
+    }
+
+    // range or single selection
+    const match = uriHunkLine.match(/H(\d*)L(\d*)L?(\d*)/);
+    if (match) {
+      if (match[3]) {
+        return (
+          match[1] === hunkIdx.toString() && lineIdx.toString() === match[2]
+        );
+      } else {
+        return uriHunkLine === "H" + hunkIdx + "L" + lineIdx;
+      }
+    }
+
+    return false;
+  }
+
+  function calcNewHash(
+    path: string,
+    hunkIdx: number,
+    lineIdx: number,
+    shift: boolean,
+  ): string {
+    // single line selection
+    if (!shift) {
+      return path + ":H" + hunkIdx + "L" + lineIdx;
+    }
+
+    if (!selectedHunkLine) {
+      return "";
+    }
+
+    // range selection
+    const match = selectedHunkLine.match(/H(\d*)L(\d*)L?(\d*)/);
+    if (match) {
+      // make sure we're on the same hunk
+      if (match[1] === hunkIdx.toString()) {
+        if (lineIdx >= parseInt(match[2])) {
+          return path + ":H" + hunkIdx + "L" + match[2] + "L" + lineIdx;
+        } else {
+          return path + ":H" + hunkIdx + "L" + lineIdx + "L" + match[2];
+        }
+      }
+    }
```

Comment:
There are three places where we parse a hunk selector. There should be only one place  where parsing happens and it should happen as early as possible.